### PR TITLE
fix: recover transient workspace bootstrap scans

### PR DESCRIPTION
**Pull request by cryppadotta**
## Thinking Path

> - Paperclip is the open source app people use to manage AI agents for work.
> - The control plane prepares task workspaces before it starts an agent.
> - Workspace preparation reads Git state so it can preserve edits and exclude private files.
> - A failed scan was treated as a non-Git folder and lost its actual failure code.
> - The resulting generic setup failure could not recover, even when the cause was temporary.
> - This pull request keeps the cause and uses the existing bounded retry schedule before provider startup.
> - Tasks can recover without human intervention, while permanent failures and exhausted retries stop with useful guidance.

## Linked Issues or Issue Description

**What happened?**

A Git scan error during managed repository preparation became `Configured repository folder is not a Git checkout`, followed by generic `setup_failed`. The agent never started. Generic recovery could not distinguish a temporary timeout from a bad workspace configuration.

**Expected behavior**

Keep the closed scan error code. Retry temporary timeouts and queue saturation under the existing shared budget. Preserve edits, exclusions, ownership, and pause gates. Stop permanent failures and exhausted retries with a specific explanation. Do not replay historical generic setup failures.

**Steps to reproduce**

1. Configure a task project with a local Git source that must be copied into its managed repositories.
2. Make the ignored-file scan exceed its timeout before the agent starts.
3. Before this fix, the snapshot returns null and the run ends as non-retryable `setup_failed`.
4. Use the disposable browser fixture in `tests/e2e/workspace-bootstrap/README.md` to inject real timeouts and test the full recovery path.

**Paperclip version or commit**

Reproduced against `4510bf7c9e2fcbeb043445850928b5dcb79908ca`.

**Deployment mode**

Built from source. The defect is in core workspace setup, not a specific model provider.

Related work: Refs #13442 (managed repository preparation), Refs #11572 (bounded Git scheduler), Refs #12997 (separate adapter startup retry work), Refs #13469 (separate terminal-workspace scan performance work).

## What Changed

- Return the non-Git fallback only for repository discovery. Propagate failed scans of a confirmed repository.
- Replace full ignored status output with an ignored-only directory listing. Preserve NUL-delimited paths and exclusions.
- Preserve typed, sanitized scan errors through workspace preparation and persist pre-provider failure details.
- Retry only timeouts and queue saturation, using the existing durable two-retry budget and issue gates. Prevent generic recovery from adding another budget.
- Show workspace-specific failure copy and actionable exhausted-recovery notices.
- Add red-green unit tests, real-database restart and retry-boundary tests, and opt-in browser acceptance fixtures with real Git subprocess timeouts.
- Document the recovery contract and browser verification procedure.

## Verification

- Red: injected scan failures returned null instead of rejecting; setup lost the timeout code; task-thread and recovery notices had generic copy.
- Green: 119 focused adapter/backend tests, 20 recovery-boundary tests, and 136 task-thread tests.
- `pnpm -r typecheck` — passed.
- `pnpm build` — passed on the final production code.
- `pnpm check:token-gates` — passed.
- The initial local `pnpm test:run` overlapped source edits and was interrupted after two late-added assertions saw pre-fix behavior; it is not counted as a green full run. A fresh final-head run passed all 229 tests across the six affected adapter/backend/UI suites. The clean latest-head CI full test matrix passed: all five general-server shards, all five serialized-server shards, and all three general-workspace shards.
- Latest-head CI also passed all three browser shards and their aggregate gate, typecheck and release registry, build, runner verification, canary dry run, policy, Docker context integrity, and security gates. Greptile: 5/5, with the review thread resolved.
- Browser: created a task in a disposable instance. A real Git timeout scheduled recovery, the next run completed through the run-scoped API without manual Retry, and Done survived reload. The deterministic process worker checked preserved source edits and excluded private files; no model calls were made.
- `WORKSPACE_BOOTSTRAP_TEST_URL=<disposable-instance-url> pnpm exec playwright test --config tests/e2e/workspace-bootstrap/playwright.config.ts` — 2 passed (3.6 minutes). The persistent case made exactly three failed attempts, never started the worker, showed the cause-specific notice, stayed stopped for another scheduler tick, and retained Blocked after reload.
- Extra red-green coverage: 50 recovery tests passed after fixing an exhausted-bootstrap classification that incorrectly implied unknown provider actions. Missing or uncertain evidence still retains the safety hold.
- Verified the documented Git executable override during repository seeding.

## Risks

- A confirmed repository scan failure now fails closed instead of falling back to directory sync. This prevents unfiltered copying but makes previously hidden errors visible.
- Temporary host problems can create up to two additional setup attempts, 30 seconds apart. Permanent scan errors do not auto-retry. Generic recovery cannot reset this budget.
- The durable retry path still enforces ownership, pause, and work eligibility. Integration tests cover restart, duplicate promotion, pause, exhaustion, and non-retryable categories.
- No schema migration, new runtime setting, new retry budget, production deployment, or historical task replay.

## Model Used

OpenAI Codex, GPT-5-based coding agent, with reasoning, repository tools, shell execution, and browser testing. The exact deployment model ID and context-window size are not exposed in this session.

## Checklist

- [x] I have included a thinking path that traces from project context to this change
- [x] I have specified the model used (with version and capability details)
- [x] I have checked ROADMAP.md and confirmed this PR does not duplicate planned core work
- [x] I have searched GitHub for duplicate or related PRs and linked them above
- [x] I have either (a) linked existing issues with `Fixes: #` / `Closes #` / `Refs #` OR (b) described the issue in-PR following the relevant issue template
- [x] I have not referenced internal/instance-local Paperclip issues or links (only public GitHub `#NNN` / `github.com/paperclipai/paperclip` URLs)
- [x] My branch name describes the change (e.g. `docs/...`, `fix/...`) and contains no internal Paperclip ticket id or instance-derived details
- [x] I have run tests locally and they pass
- [x] I have added or updated tests where applicable
- [x] I have updated relevant documentation to reflect my changes
- [x] I have considered and documented any risks above
- [x] All Paperclip CI gates are green
- [x] Greptile is 5/5 with no open P2s, recommendations, or follow-ups
- [x] I will address all Greptile and reviewer comments before requesting merge

### CODE REVIEW

**Comment by cryppadotta** [comment]
@greptile review

File: tests/e2e/workspace-bootstrap/seed.mjs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  const api = async (route, method = "GET", body) => {
+    const response = await fetch(`${base}/api${route}`, { method, headers: { "Content-Type": "application/json" }, body: body ? JSON.stringify(body) : undefined });
+    assert(response.ok, `${response.status}: ${await response.clone().text()}`);
+    return response.json();
+  };
+  const health = await api("/health");
+  assert.equal(health.deploymentMode, "local_trusted");
+  const [company] = await api("/companies");
+  assert.equal(company.name, "Workspace Recovery QA");
+  const source = await mkdtemp(path.join(os.tmpdir(), "workspace-bootstrap-source-"));
```
[the remark this answers is not in the document]
Comment (cryppadotta) [reply tests/e2e/workspace-bootstrap/seed.mjs:20]: Fixed in 1845ed53e. Repository seeding now uses BOOTSTRAP_REAL_GIT, as the timeout shim already did. Verified seeding with a non-default Git executable. The acceptance timeout also accounts for scheduler tick delay.

**Comment by cryppadotta** [comment]
@greptile review
